File: L/1139-largest-1-bordered-square/main.go
package main

import "fmt"

func largest1BorderedSquare(grid [][]int) int {
	buf := make([][][]int, len(grid))
	for i := 0; i < len(grid); i++ {
		buf[i] = make([][]int, len(grid[0]))
	}
	max := -1
	for i := len(grid) - 1; i >= 0; i-- {
		for j := len(grid[0]) - 1; j >= 0; j-- {
			if grid[i][j] == 0 {
				buf[i][j] = []int{0, 0}
				continue
			}

			r, c := 1, 1
			if j < len(grid[0])-1 {
				r = buf[i][j+1][0] + 1
			}
			if i < len(grid)-1 {
				c = buf[i+1][j][1] + 1
			}
			buf[i][j] = []int{r, c}
			cnt := r
			if cnt > c {
				cnt = c
			}
			for s := cnt - 1; s >= 0; s-- {
				if buf[i][j+s][1] >= s+1 && buf[i+s][j][0] >= s+1 && s > max {
					max = s
				}
			}
		}
	}

	if max == -1 {
		return 0
	}
	return (max + 1) * (max + 1)
}

func main() {
	fmt.Println()
}

File: L/1139-largest-1-bordered-square/main.rs
struct Solution {}

impl Solution {
  pub fn largest1_bordered_square(grid: Vec<Vec<i32>>) -> i32 {
    let mut buf: Vec<Vec<[usize; 2]>> = Vec::new();

    (0..grid.len()).for_each(|r| {
      let mut col: Vec<[usize; 2]> = Vec::new();
      (0..grid[0].len()).for_each(|c| {
        if grid[r][c] == 0 {
          col.push([0, 0]);
          return;
        }
        let mut v: [usize; 2] = [0, 0];
        match (r, c) {
          (0, 0) => {
            v = [grid[r][c] as usize, grid[r][c] as usize];
          }
          (0, _) => {
            v = [col[c - 1][0] + 1, 1];
          }
          (_, 0) => {
            v = [1, buf[r - 1][c][1] + 1];
          }
          (_, _) => {
            v = [col[c - 1][0] + 1, buf[r - 1][c][1] + 1];
          }
        }
        col.push(v);
      });
      buf.push(col);
    });
    let mut ret: usize = 0;
    (0..grid.len()).for_each(|r| {
      (0..grid[0].len()).for_each(|c| {
        let mut size: usize = r + 1;
        if size > c + 1 {
          size = c + 1;
        }
        if size > buf[r][c][0] {
          size = buf[r][c][0];
        }
        if size > buf[r][c][1] {
          size = buf[r][c][1];
        }

        let mut cret: usize = 0;
        (0..size).rev().for_each(|l| {
          if buf[r][c - l][1] >= l + 1 && buf[r - l][c][0] >= l + 1 && l + 1 > cret {
            cret = l + 1;
          }
        });
        if cret > ret {
          ret = cret;
        }
      });
    });
    (ret * ret) as i32
  }
}
